web/packages/teleport/src/AppLauncher/AppLauncher.tsx

```diff
@@ -76,7 +54,33 @@ export function AppLauncher() {
         }
       }
 
-      window.location.replace(`https://${fqdn}${port}${path}`);
+      // Let the target app know of a new auth exchange.
+      const stateToken = queryParams.get('state');
+      if (!stateToken) {
+        initiateNewAuthExchange({ fqdn, port, path, params });
+        return;
+      }
+
+      // Continue the auth exchange.
+
+      if (params.arn) {
+        params.arn = decodeURIComponent(params.arn);
+      }
+      const session = await service.createAppSession(params);
+
+      // Set all the fields expected by server to validate request.
+      const url = getXTeleportAuthUrl({ fqdn, port });
+      url.searchParams.set('state', stateToken);
+      url.searchParams.set('subject', session.subjectCookieValue);
+      url.hash = `#value=${session.cookieValue}`;
```

Comment:
Are we sure that whatever there is in the `cookieValue` variable doesn't need to be URL-escaped?

Reply:
it should be safe since the `cookieValue` is the [session id](https://github.com/gravitational/teleport/blob/95aa537f681555f460bdee212f6c96a3a2c901c9/lib/auth/sessions.go#L100), which is a hex string